src/mbgl/renderer/painter_fill.cpp

```diff
@@ -90,6 +90,26 @@ void Painter::renderFill(FillBucket& bucket, const FillLayer& layer, const TileI
             patternShader->u_patternmatrix_a = patternMatrixA;
             patternShader->u_patternmatrix_b = patternMatrixB;
 
+            std::array<int, 2> imageSizeScaledA = {{
+                (int)((*posA).size[0] * properties.pattern.value.fromScale),
+                (int)((*posA).size[1] * properties.pattern.value.fromScale)
+            }};
+            std::array<int, 2> imageSizeScaledB = {{
+                (int)((*posB).size[0] * properties.pattern.value.toScale),
+                (int)((*posB).size[1] * properties.pattern.value.toScale)
+            }};
+
+            int tileSize = 512;
+
+            float offsetAx = ((tileSize % imageSizeScaledA[0]) * id.x) / (float)imageSizeScaledA[0];
```

Comment:
Is `w` not available on the native side? https://github.com/mapbox/mapbox-gl-js/pull/1908/files#diff-a406b8ffc7f2d9f303442aecebb9a2bdR164